File: packages/sdk-java/harness-sdk-core/src/main/java/com/harness/core/StuckDetector.java
package com.harness.core;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.harness.types.Message;

/**
 * Detect when agent is stuck using semantic similarity.
 *
 * Uses embedding model to detect repetitive outputs that indicate
 * the agent is not making progress.
 *
 * Example:
 * <pre>
 * StuckDetectorConfig config = StuckDetectorConfig.builder()
 *     .enableSemantic(true)
 *     .similarityThreshold(0.92)
 *     .consecutiveRounds(3)
 *     .build();
 *
 * StuckDetector detector = new StuckDetector(config, embeddingModel);
 *
 * // Check after each tool execution
 * StuckDetectionResult result = detector.check(sessionId, session.messages(), iteration);
 * if (result.isStuck()) {
 *     // Inject feedback or terminate
 * }
 * </pre>
 */
public class StuckDetector {

    private static final Logger logger = LoggerFactory.getLogger(StuckDetector.class);

    private final StuckDetectorConfig config;
    private final EmbeddingModel embeddingModel;

    // Per-session state
    private final Map<String, Deque<float[]>> windows = new HashMap<>();
    private final Map<String, Integer> consecutive = new HashMap<>();

    // Embedding cache (text_hash -> embedding)
    private final Map<String, float[]> cache = new LinkedHashMap<>();
    private static final int MAX_CACHE_SIZE = 10000;

    public StuckDetector() {
        this(StuckDetectorConfig.defaults(), null);
    }

    public StuckDetector(StuckDetectorConfig config) {
        this(config, null);
    }

    public StuckDetector(StuckDetectorConfig config, EmbeddingModel embeddingModel) {
        this.config = config;
        this.embeddingModel = embeddingModel;
    }

    /**
     * Check if agent is stuck using semantic similarity.
     *
     * @param sessionId Session identifier
     * @param messages Recent messages (tool outputs, assistant responses)
     * @param iteration Current iteration number
     * @return StuckDetectionResult with detection outcome
     */
    public StuckDetectionResult check(String sessionId, List<Message> messages, int iteration) {
        // Check if semantic detection is enabled
        if (!config.enableSemantic()) {
            return StuckDetectionResult.notStuck("semantic_disabled");
        }

        // Check if embedding model is available
        if (embeddingModel == null || !embeddingModel.isAvailable()) {
            return StuckDetectionResult.notStuck("model_unavailable");
        }

        // Extract candidate texts
        List<String> texts = extractTexts(messages);
        if (texts.isEmpty()) {
            return StuckDetectionResult.notStuck("no_candidates", Map.of("messageCount", messages.size()));
        }

        // Combine texts for embedding (one embedding per iteration)
        String combinedText = String.join("\n", texts.subList(Math.max(0, texts.size() - 3), texts.size()));

        // Get embedding
        float[] embedding = getEmbedding(combinedText);
        if (embedding == null) {
            return StuckDetectionResult.notStuck("embedding_failed");
        }

        // Get or create window
        Deque<float[]> window = windows.computeIfAbsent(
            sessionId,
            k -> new ArrayDeque<>(config.windowSize())
        );

        // Calculate max similarity against window
        double maxSim = 0.0;
        List<Double> similarities = new ArrayList<>();
        for (float[] prevEmb : window) {
            double sim = cosineSimilarity(embedding, prevEmb);
            similarities.add(sim);
            maxSim = Math.max(maxSim, sim);
        }

        // Update window
        window.addLast(embedding);
        if (window.size() > config.windowSize()) {
            window.removeFirst();
        }

        // Update consecutive count
        int consecutiveCount = this.consecutive.getOrDefault(sessionId, 0);
        if (maxSim >= config.similarityThreshold()) {
            consecutiveCount++;
        } else {
            consecutiveCount = 0;
        }
        this.consecutive.put(sessionId, consecutiveCount);

        // Determine if stuck
        boolean isStuck = consecutiveCount >= config.consecutiveRounds();

        double avgSim = similarities.isEmpty() ? 0.0 : similarities.stream().mapToDouble(d -> d).average().orElse(0.0);

        Map<String, Object> details = new HashMap<>();
        details.put("maxSimilarity", maxSim);
        details.put("avgSimilarity", avgSim);
        details.put("windowSize", window.size());
        details.put("consecutive", consecutiveCount);
        details.put("textPreview", combinedText.length() > 200 ? combinedText.substring(0, 200) : combinedText);

        if (isStuck) {
            return StuckDetectionResult.stuck("semantic_repeat", maxSim, consecutiveCount, details);
        }

        return new StuckDetectionResult(false, "no_stuck", maxSim, consecutiveCount, details);
    }

    /**
     * Clear session state.
     *
     * Call when session ends or after feedback injection.
     */
    public void clearSession(String sessionId) {
        windows.remove(sessionId);
        consecutive.remove(sessionId);
    }

    /**
     * Reset all state.
     */
    public void reset() {
        windows.clear();
        consecutive.clear();
        cache.clear();
    }

    // === Private Methods ===

    private List<String> extractTexts(List<Message> messages) {
        List<String> candidates = new ArrayList<>();
        for (Message msg : messages) {
            if ("tool".equals(msg.role()) || "assistant".equals(msg.role())) {
                String content = msg.contentAsString();
                if (content != null) {
                    String normalized = normalizeText(content);
                    if (normalized.length() >= config.minChars()) {
                        candidates.add(normalized);
                    }
                }
            }
        }
        return candidates;
    }

    private float[] getEmbedding(String text) {
        String hash = textHash(text);

        // Check cache
        if (cache.containsKey(hash)) {
            return cache.get(hash);
        }

        try {
            float[] embedding = embeddingModel.embed(text);
            if (embedding != null) {
                // Cache it
                if (cache.size() >= MAX_CACHE_SIZE) {
                    // Simple LRU: remove first entry
                    String firstKey = cache.keySet().iterator().next();
                    cache.remove(firstKey);
                }
                cache.put(hash, embedding);
            }
            return embedding;
        } catch (Exception e) {
            logger.warn("StuckDetector: Embedding failed: {}", e.getMessage());
            return null;
        }
    }

    private static String normalizeText(String s) {
        return s.trim().replaceAll("\\s+", " ");
    }

    private static String textHash(String s) {
        try {
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] digest = md.digest(s.getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder();
            for (byte b : digest) {
                sb.append(String.format("%02x", b));
            }
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            // Fallback to simple hash
            return String.valueOf(s.hashCode());
        }
    }

    private static double cosineSimilarity(float[] a, float[] b) {
        if (a == null || b == null || a.length != b.length) {
            return 0.0;
        }

        double dotProduct = 0.0;
        double normA = 0.0;
        double normB = 0.0;

        for (int i = 0; i < a.length; i++) {
            dotProduct += a[i] * b[i];
            normA += a[i] * a[i];
            normB += b[i] * b[i];
        }

        if (normA == 0 || normB == 0) {
            return 0.0;
        }

        return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
    }
}

File: packages/sdk/src/harness/core/stuck_detector.py
"""
Stuck Detector - Detect when agent is stuck in non-progressing state.

Implements multiple detection strategies:
1. Empty/error detection (zero-cost, always available)
2. Semantic similarity detection (requires embedding model)

The semantic detection uses bge-small-zh-v1.5 by default, optimized for Chinese text.
"""

from __future__ import annotations

import asyncio
import hashlib
import logging
from collections import deque
from dataclasses import dataclass, field
from typing import TYPE_CHECKING

import numpy as np

if TYPE_CHECKING:
    from harness.types import Message, Session

logger = logging.getLogger(__name__)

# Default embedding model (Chinese optimized)
DEFAULT_EMBEDDING_MODEL = "bge-small-zh-v1.5"


# =============================================================================
# Configuration
# =============================================================================

@dataclass
class StuckDetectorConfig:
    """
    Configuration for stuck detection.

    Attributes:
        enable_semantic: Enable semantic similarity detection
        similarity_threshold: Cosine similarity threshold (0.0-1.0)
        consecutive_rounds: Consecutive similar rounds to trigger stuck
        window_size: Number of recent embeddings to compare against
        min_chars: Minimum characters for embedding (shorter texts are skipped)
    """
    # Core switch
    enable_semantic: bool = False

    # Semantic detection thresholds
    similarity_threshold: float = 0.92
    consecutive_rounds: int = 3
    window_size: int = 6
    min_chars: int = 30


@dataclass
class StuckDetectionResult:
    """
    Result from stuck detection.

    Attributes:
        is_stuck: Whether agent is stuck
        reason: "empty", "error", "semantic_repeat", "no_stuck", "model_unavailable"
        similarity: Max similarity score (if semantic detection ran)
        consecutive_count: Current consecutive similar count
        details: Additional diagnostic information
    """
    is_stuck: bool
    reason: str
    similarity: float | None = None
    consecutive_count: int = 0
    details: dict = field(default_factory=dict)


# =============================================================================
# Utility Functions
# =============================================================================

def _normalize_text(s: str) -> str:
    """Normalize text for embedding."""
    return " ".join(s.strip().split())


def _text_hash(s: str) -> str:
    """Generate hash for text (for caching)."""
    return hashlib.md5(s.encode("utf-8")).hexdigest()


def _cosine_similarity(a: np.ndarray, b: np.ndarray) -> float:
    """Calculate cosine similarity between two vectors."""
    if a is None or b is None:
        return 0.0
    norm_a = np.linalg.norm(a)
    norm_b = np.linalg.norm(b)
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return float(np.dot(a, b) / (norm_a * norm_b))


# =============================================================================
# StuckDetector
# =============================================================================

class StuckDetector:
    """
    Detect when agent is stuck using semantic similarity.

    Uses embedding model to detect repetitive outputs that indicate
    the agent is not making progress.

    Example:
        config = StuckDetectorConfig(enable_semantic=True)
        detector = StuckDetector(config)

        # Check after each tool execution
        result = await detector.check(session.id, session.messages, iteration)
        if result.is_stuck:
            # Inject feedback or terminate
            pass
    """

    def __init__(self, config: StuckDetectorConfig | None = None):
        self.config = config or StuckDetectorConfig()

        # Lazy-loaded model
        self._model = None
        self._model_unavailable = False
        self._executor = None

        # Per-session state
        self._windows: dict[str, deque] = {}  # session_id -> deque[embedding]
        self._consecutive: dict[str, int] = {}  # session_id -> consecutive count

        # Embedding cache (text_hash -> embedding)
        self._cache: dict[str, np.ndarray] = {}
        self._cache_order: deque = deque(maxlen=10000)  # LRU eviction

    def _load_model(self) -> bool:
        """
        Lazy load embedding model.

        Returns:
            True if model loaded successfully, False otherwise
        """
        if self._model is not None:
            return True
        if self._model_unavailable:
            return False  # Avoid repeated attempts

        try:
            from sentence_transformers import SentenceTransformer

            self._model = SentenceTransformer(DEFAULT_EMBEDDING_MODEL)
            logger.info(f"StuckDetector: Loaded {DEFAULT_EMBEDDING_MODEL}")
            return True
        except ImportError:
            logger.warning(
                "StuckDetector: sentence-transformers not installed. "
                "Install with: pip install harness-sdk[stuck]"
            )
            self._model_unavailable = True
            return False
        except Exception as e:
            logger.warning(f"StuckDetector: Failed to load model: {e}")
            self._model_unavailable = True
            return False

    async def _get_embedding(self, text: str) -> np.ndarray | None:
        """
        Get embedding for text, using cache if available.

        Args:
            text: Text to embed

        Returns:
            Embedding vector or None if failed
        """
        h = _text_hash(text)

        # Check cache
        if h in self._cache:
            return self._cache[h]

        # Load model if needed
        if not self._load_model():
            return None

        try:
            # Run in executor to avoid blocking
            loop = asyncio.get_running_loop()
            if self._executor is None:
                from concurrent.futures import ThreadPoolExecutor
                self._executor = ThreadPoolExecutor(max_workers=1)

            embeddings = await loop.run_in_executor(
                self._executor,
                lambda: self._model.encode([text], convert_to_numpy=True)
            )

            emb = np.asarray(embeddings[0], dtype=np.float32)

            # Cache it
            self._cache[h] = emb
            self._cache_order.append(h)

            return emb
        except Exception as e:
            logger.warning(f"StuckDetector: Embedding failed: {e}")
            return None

    def _extract_texts(self, messages: list[Message]) -> list[str]:
        """
        Extract candidate texts from messages for embedding.

        Prioritizes tool outputs and assistant responses.

        Args:
            messages: List of messages

        Returns:
            List of texts to embed
        """
        candidates = []
        for msg in messages:
            if msg.role in ("tool", "assistant"):
                content = msg.content
                if isinstance(content, str):
                    text = _normalize_text(content)
                elif isinstance(content, list):
                    # Handle structured content
                    text = _normalize_text(str(content))
                else:
                    text = _normalize_text(str(content))

                if len(text) >= self.config.min_chars:
                    candidates.append(text)

        return candidates

    async def check(
        self,
        session_id: str,
        messages: list[Message],
        iteration: int,
    ) -> StuckDetectionResult:
        """
        Check if agent is stuck using semantic similarity.

        Args:
            session_id: Session identifier
            messages: Recent messages (tool outputs, assistant responses)
            iteration: Current iteration number

        Returns:
            StuckDetectionResult with detection outcome
        """
        # Check if semantic detection is enabled
        if not self.config.enable_semantic:
            return StuckDetectionResult(
                is_stuck=False,
                reason="semantic_disabled",
            )

        # Extract candidate texts
        texts = self._extract_texts(messages)
        if not texts:
            return StuckDetectionResult(
                is_stuck=False,
                reason="no_candidates",
                details={"message_count": len(messages)},
            )

        # Combine texts for embedding (one embedding per iteration)
        combined_text = "\n".join(texts[-3:])  # Last 3 messages

        # Get embedding
        embedding = await self._get_embedding(combined_text)
        if embedding is None:
            return StuckDetectionResult(
                is_stuck=False,
                reason="model_unavailable",
            )

        # Get or create window
        window = self._windows.setdefault(
            session_id,
            deque(maxlen=self.config.window_size)
        )

        # Calculate max similarity against window
        max_sim = 0.0
        similarities = []
        for prev_emb in window:
            sim = _cosine_similarity(embedding, prev_emb)
            similarities.append(sim)
            max_sim = max(max_sim, sim)

        # Update window
        window.append(embedding)

        # Update consecutive count
        consecutive = self._consecutive.get(session_id, 0)
        if max_sim >= self.config.similarity_threshold:
            consecutive += 1
        else:
            consecutive = 0
        self._consecutive[session_id] = consecutive

        # Determine if stuck
        is_stuck = consecutive >= self.config.consecutive_rounds

        details = {
            "max_similarity": max_sim,
            "avg_similarity": float(np.mean(similarities)) if similarities else 0.0,
            "window_size": len(window),
            "consecutive": consecutive,
            "text_preview": combined_text[:200],
        }

        if is_stuck:
            return StuckDetectionResult(
                is_stuck=True,
                reason="semantic_repeat",
                similarity=max_sim,
                consecutive_count=consecutive,
                details=details,
            )

        return StuckDetectionResult(
            is_stuck=False,
            reason="no_stuck",
            similarity=max_sim,
            consecutive_count=consecutive,
            details=details,
        )

    def clear_session(self, session_id: str) -> None:
        """
        Clear session state.

        Call when session ends or after feedback injection.

        Args:
            session_id: Session to clear
        """
        self._windows.pop(session_id, None)
        self._consecutive.pop(session_id, None)

    def reset(self) -> None:
        """Reset all state."""
        self._windows.clear()
        self._consecutive.clear()
        self._cache.clear()
        self._cache_order.clear()
